pkg/agent/agent_test.go

```diff
@@ -0,0 +1,139 @@
+package pv_monitor_agent
+
+import (
+	"os"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/kubernetes-csi/external-health-monitor/pkg/mock"
+)
+
+func Test_AbnormalVolume(t *testing.T) {
+	abnormalVolume := &mock.MockVolume{
+		CSIVolume: &mock.CSIVolume{
+			Volume: &csi.Volume{
+				VolumeId: "abnormalVolume1",
+			},
+			Condition: &csi.VolumeCondition{
+				Abnormal: true,
+				Message:  "Volume not found",
+			},
+		},
+		NativeVolume:      mock.CreatePV(2, "pvc", "pv", mock.DefaultNS, "abnormalVolume1", "pvcuid", &mock.FSVolumeMode, v1.VolumeBound),
+		NativeVolumeClaim: mock.CreatePVC(1, 2, "pvc", "pvcuid", mock.DefaultNS, "pv", v1.ClaimBound),
+	}
+
+	pod := &mock.MockPod{
+		NativePod: mock.CreatePod("pod", mock.DefaultNS, "pv", "pvc", "node", "uid", false),
+	}
+
+	testCase := &testCase{
+		name: "abnormal_volume_case1",
+		fakeNativeObjects: &fakeNativeObjects{
+			MockVolume: abnormalVolume,
+			MockPod:    pod,
+		},
+		wantEvent: true,
+	}
+
+	os.Setenv("NODE_NAME", "node")
+	runTest(t, testCase)
+}
+
+func Test_AbnormalVolumeVolumePending(t *testing.T) {
+	abnormalVolume := &mock.MockVolume{
+		CSIVolume: &mock.CSIVolume{
+			Volume: &csi.Volume{
+				VolumeId: "abnormalVolume1",
+			},
+			Condition: &csi.VolumeCondition{
+				Abnormal: true,
+				Message:  "Volume not found",
+			},
+		},
+		NativeVolume:      mock.CreatePV(2, "pvc", "pv", mock.DefaultNS, "abnormalVolume1", "pvcuid", &mock.FSVolumeMode, v1.VolumePending),
```

Comment:
We are only checking volumes that are bound.  If volumes are not bound, there should already be events generated by the existing controllers.  Also the Message says "Volume not found" which does not match "Pending".
So change v1.VolumePending to v1.Bound

Rename "Test_AbnormalVolumeVolumePending" to "Test_AbnormalVolumeIOError".  Change Message from "Volume not found" to "IO error".

Reply:
Only test different error message looks like not necessary. Let me remove this test case